pkg/broker/ingress.go

```diff
@@ -0,0 +1,101 @@
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package broker
+
+import (
+	"context"
+	"strings"
+
+	"github.com/cloudevents/sdk-go/pkg/cloudevents"
+	eventingv1alpha1 "github.com/knative/eventing/pkg/apis/eventing/v1alpha1"
+	"go.uber.org/zap"
+	apierrs "k8s.io/apimachinery/pkg/api/errors"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// Policy is a helper struct to determine if the CloudEvents received pass the Broker's policy.
+type Policy struct {
+	logger    *zap.SugaredLogger
+	client    client.Client
+	namespace string
+	broker    string
+	spec      *eventingv1alpha1.BrokerPolicySpec
+}
+
+// NewPolicy creates a Policy for a particular Broker.
+func NewPolicy(logger *zap.Logger, client client.Client, allowAny bool, namespace, broker string) *Policy {
+	return &Policy{
+		logger:    logger.Sugar(),
+		client:    client,
+		namespace: namespace,
+		broker:    broker,
+		spec: &eventingv1alpha1.BrokerPolicySpec{
+			AllowAny: allowAny,
+		},
+	}
+}
+
+// AllowEvent filters events based on the configured policy.
+func (p *Policy) AllowEvent(ctx context.Context, event cloudevents.Event) bool {
+	// 1. If allowAny is set to true, then all events are allowed to enter the mesh.
+	// 2. If allowAny is set to false, then the event is only accepted if it's already in the Broker's registry.
+	if p.spec.AllowAny {
+		p.logger.Debugf("EventType (type:'%s', source:'%s', schema:'%s') received, Accept", event.Type(), event.Source(), event.SchemaURL())
+		return true
+	}
+	return p.isRegistered(ctx, event)
+}
+
+// isRegistered returns whether the EventType corresponding to the CloudEvent 'event' is available in the Registry.
+// Use with caution as this method can affect the data plane throughput.
+func (p *Policy) isRegistered(ctx context.Context, event cloudevents.Event) bool {
+	_, err := p.getEventType(ctx, event)
+	if apierrs.IsNotFound(err) {
+		p.logger.Debugf("EventType (type:'%s', source:'%s', schema:'%s') not found, Reject", event.Type(), event.Source(), event.SchemaURL())
+		return false
+	} else if err != nil {
+		p.logger.Errorf("Error retrieving EventType (type:'%s', source:'%s', schema:'%s'), Reject: %v", event.Type(), event.Source(), event.SchemaURL(), err)
+		return false
+	}
+	p.logger.Debugf("EventType (type:'%s', source:'%s', schema:'%s') is registered, Accept", event.Type(), event.Source(), event.SchemaURL())
+	return true
+}
+
+// getEventType retrieves the EventType from the Registry for the given CloudEvent 'event'.
+// If it is not found, it returns an error.
+func (p *Policy) getEventType(ctx context.Context, event cloudevents.Event) (*eventingv1alpha1.EventType, error) {
```

Comment:
Does this client always goes to the API server? If so, that's bad...
What is a better way of doing this to query just the cached ones in memory? 
And whenever a new EventType is added to the API server, then the cache is invalidated.